src/search.c

```diff
@@ -40,6 +41,121 @@ bool init = false;
 static char search_type = '0';
 static char scope = '0';
 
+struct bundle_result {
+	char bundle_name[256];
+	long size;
+	double score;
+	struct list *files;
+};
+
+struct file_result {
+	char filename[PATH_MAX+1];
+	double score;
+};
+
+static struct list *results;
+
+
+static void add_bundle_file_result(char *bundlename, char *filename, double score, struct manifest *man) 
+{
+	struct bundle_result *bundle = NULL;
+	struct file_result *file;
+	struct list *ptr;
+
+	ptr = results;
+	while (ptr) {
+		struct bundle_result *item;
+
+		item = ptr->data;
+		if (strcmp(item->bundle_name, bundlename) == 0) {
+			bundle = item;
+			break;
+		}
+		ptr = ptr->next;
+	}
+	if (!bundle) {
+		bundle = calloc(sizeof(struct bundle_result), 1);
+		results = list_append_data(results, bundle);
+		strncpy(bundle->bundle_name, bundlename, 255);
+		bundle->size = man->contentsize / 1024 / 1204;
+		bundle->score = -0.1 * bundle->size;
+	}
+
+	file = calloc(sizeof(struct file_result), 1);
+	strncpy(file->filename, filename, PATH_MAX);
+	bundle->files = list_append_data(bundle->files, file);
+	file->score = score;
+	bundle->score += score;
+}
+
+static int bundle_cmp(const void *a, const void *b)
+{
+	const struct bundle_result *A, *B;
+	A = (struct bundle_result *)a;
+	B = (struct bundle_result *)b;
+	if (A->score > B->score)
+		return -1;
+	if (A->score < B->score)
+		return 1;
+	return 0;
+	
+}
+
+static int file_cmp(const void *a, const void *b)
+{
+	const struct file_result *A, *B;
+	A = (struct file_result *)a;
+	B = (struct file_result *)b;
+	if (A->score > B->score)
+		return -1;
+	if (A->score < B->score)
+		return 1;
+	return 0;
+	
+}
+static void sort_results(void)
+{
+	struct list *ptr;
+	struct bundle_result *b;
+
+	results = list_sort(results, bundle_cmp);
+
+	ptr = results;
+	while (ptr) {
+		b = ptr->data;
+		ptr = ptr->next;
+
+		b->files = list_sort(b->files, file_cmp);
+	}
+}
+
+static void print_final_results(void)
+{
+	struct bundle_result *b;
+	struct list *ptr;
+	int counter = 0;
+
+	ptr = results;
+	while (ptr && counter < 5) {
```

Comment:
Why not a more idiomatic for loop

    for (counter=0, ptr=results;
          ptr && counter < 5;
          ptr=ptr->next, counter++) {